### DISCUSSION TITLE
Mysql gone away for Idle connection listener for long running runtime (not solved)

### DISCUSSION THREAD

**Discussion by abhi-nav** [symfony/symfony#57217] [Q&A] [answered]
### Symfony version(s) affected

7.1

### Description

I am using symfony7.1 and runtime frankenPHP in worker mode, I am still getting Mysql has gone away Issue after multiple attempts the connection gets connected. 

this was marked solved in this commit
f7cc44e9ac0b560098c726c713d6c49f34b29658

### How to reproduce

After running the application, wait till the wait_timeout from mysql, the first error message will be 'Mysql has gone away error'
![image](https://github.com/symfony/symfony/assets/6917575/e31b7dca-6485-4336-a922-09b5fe2cbcbb)

### Additional Context

Please let me know if i have missed something during the setup.

**Comment by girinabin** [comment]
@abhi-nav i am also getting same.after multiple retry it gets connection.
https://github.com/symfony/symfony/pull/53214
this is the original issue.They said it was fixed in 7.1 but i am still getting that mysql has gone way error
@dunglas @alli83

**Comment by alli83** [comment]
from what duration, for example do you lose the SQL Connection? Also did you leave the idle_connection_ttl as its default configuration?

**Reply by abhi-nav** [reply]
yes the idle_connection_ttl is default. You are right we have also seen that we lose sql connection before wait_timeout in 7.1.

we had our custom code in place before this patch (inspired from roadrunner) (symfony 7.1) in version 6.4 if remained in idle state we get **mysql gone away error** exactly after default wait_timeout (8hr) . The connection is re-stablished after first error attempt. 

```
class DoctrineConnectionPingerMiddleware implements EventSubscriberInterface
{
    public function __construct(private EntityManagerInterface $em)
    {
    }

    #[\Override] public static function getSubscribedEvents(): array
    {
        return [
            KernelEvents::REQUEST => 'onKernelRequest',
        ];
    }

    public function onKernelRequest(): void
    {
        if (
            $this->em->getConnection()->isConnected() &&
            false === $this->ping($this->em->getConnection())
        ) {
            $this->em->getConnection()->close();
            $this->em->getConnection()->connect();
        }
    }

    private function ping(Connection $con): bool
    {
        try {
            $con->executeQuery($con->getDatabasePlatform()->getDummySelectSQL());
            return true;
        } catch (\Exception) {
            return false;
        }
    }
}
```

But in consumer worker we never have a problem after this simple configuration in messenger.yaml
```
framework:
    messenger:
        # Uncomment this (and the failed transport below) to send failed messages to this transport for later handling.
        # failure_transport: failed
        buses:   
            default:
                middleware: 
                    - doctrine_ping_connection
                    - doctrine_close_connection 
```
Why cant we use similar logic like these middleware  🤔

@alli83 could you replicate this at your end ? or am i missing any config

**Comment by alli83** [comment]
I'm going to try to reproduce it.

**Comment by alli83** [comment]
I have set up a project where I force the MySQL` 'wait_timeout'` to 1 and set` idle_connection_ttl` configuration to 1 and in this case, I can't reproduce your problem; I don't lose the connection between queries. The listener works well.

To better understand your problem, could you provide more details about your implementation? also, is it indeed symfony 7.1 and doctrine bundle 2.13?

**Reply by girinabin** [reply]
@alli83
 #### _I am using symfony 7.1 with franken php in worker mode_
#### My composer.json
```composer.json
  "require": {
    "php": ">=8.3",
    "ext-ctype": "*",
    "ext-curl": "*",
    "ext-fileinfo": "*",
    "ext-iconv": "*",
    "ext-imagick": "*",
    "ext-json": "*",
    "ext-pdo": "*",
    "aws/aws-sdk-php": "^3.300",
    "beberlei/assert": "^3.3.1",
    "composer/package-versions-deprecated": "1.11.99.3",
    "cuyz/valinor": "^1.9",
    "doctrine/dbal": "^4.0",
    "doctrine/doctrine-bundle": "^2.12",
    "doctrine/doctrine-fixtures-bundle": "^3.4",
    "doctrine/doctrine-migrations-bundle": "^3.3",
    "doctrine/orm": "^3.2",
    "koco/messenger-kafka": "^0.18.0",
    "nelmio/cors-bundle": "^2.1",
    "ramsey/uuid": "^4.2.1",
    "ramsey/uuid-doctrine": "^2.0",
    "runtime/frankenphp-symfony": "^0.2.0",
    "spomky-labs/otphp": "^11.1",
    "symfony/console": "7.1.*",
    "symfony/doctrine-messenger": "7.1.*",
    "symfony/dotenv": "7.1.*",
    "symfony/expression-language": "7.1.*",
    "symfony/flex": "^2.3.1",
    "symfony/framework-bundle": "7.1.*",
    "symfony/http-client": "7.1.*",
    "symfony/lock": "7.1.*",
    "symfony/mailer": "7.1.*",
    "symfony/messenger": "7.1.*",
    "symfony/monolog-bundle": "^3.7",
    "symfony/runtime": "7.1.*",
    "symfony/security-bundle": "7.1.*",
    "symfony/serializer": "7.1.*",
    "symfony/validator": "7.1.*",
    "symfony/yaml": "7.1.*",
    "web-token/jwt-library": "^3.3"
  },
  "require-dev": {
    "friendsofphp/php-cs-fixer": "^3.1",
    "symfony/stopwatch": "7.1.*"
  },
```
### doctrine/doctrine-bundle stable version available is 2.12.0(2.13 not stable)
### _To reproduce follow this steps:_
#### 1.Must run in worker mode
``` Caddyfile
    worker {
		file ./public/index.php
		num 3
		env APP_RUNTIME Runtime\FrankenPhpSymfony\Runtime
	}
``` 
#### 2. Doctrine.yml config
```yml
doctrine:
  dbal:
    url: '%env(resolve:CENTRAL_DATABASE_URL)%'
    options:
      1002: 'SET wait_timeout = 1'
```
#### 3. Default controller
```php
<?php

namespace App\Controller;

use Doctrine\DBAL\Connection;
use Doctrine\ORM\EntityManagerInterface;
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;

class DefaultController extends AbstractController
{
    public function __construct(
        private EntityManagerInterface $em
    )
    {
    }

    public function index()
    {
        var_dump('is connected?', $this->em->getConnection()->isConnected());
        var_dump('ping', $this->ping($this->em->getConnection()));
        $stmt = $this->em->getConnection()->prepare('select * from ROLES');
        $data = $stmt->executeQuery()->fetchAllAssociative();

        return $this->json($data);
    }

    private function ping(Connection $con): bool
    {
        try {
            $con->executeQuery($con->getDatabasePlatform()->getDummySelectSQL());
            return true;
        } catch (\Exception) {
            return false;
        }
    }

}

```
#### _Since worker mode thread is 3 after 3 attempts only mysql gone issue will go_.

**Comment by alli83** [comment, marked as answer]
it has been merged in DoctrineBundle 2.13

**Reply by abhi-nav** [reply]
@alli83 Thank you for trying to replicate this issue. This fix is really crucial for our project 
@girinabin  Thank you for providing step by step guide.

Somehow the code is dependent to our APP_ENV. Even the code is running in worker mode. APP_ENV=dev was giving us "gone away" issue while APP_ENV=prod is not giving us any issue. Will look more into this now. 

Anyway Alia your code is working FINE   ✅ 😍 Thanks again for the awesome merge. 

Will get back to you if we face any issues in future.

Have a good day.

**Reply by girinabin** [reply]
@abhi-nav 
after using 
`  "doctrine/doctrine-bundle": "^2.13.x-dev",
`
still facing mysql gone way issue

**Reply by girinabin** [reply]
@alli83 @abhi-nav 
I check my APP_ENV it was dev
when i make APP_ENV = prod 
issue was gone.
there is no mysql gone way issue after i changed my APP_ENV to prod.No need to update to doctrine-bundle 2.13, issue was solved even in doctrine-bundle 2.12 and  symfony 7.1

**Comment by girinabin** [comment]
@alli83 i still get one new issue:
I'm running FrankenPHP in worker mode with Doctrine ORM. I encountered an issue where, after manually closing the database connection (by bringing down the MySQL service with docker-compose down and then bringing it back up with docker-compose up), some requests start failing with a "There is no active transaction" exception. Other requests pass successfully, but the error continues for 4-5 requests before stabilizing.

i am running
"doctrine/doctrine-bundle": "^2.13",
symfony: 7.1
php 8.3

output log:
{"message":"Beginning transaction","context":{},"level":100,"level_name":"DEBUG","channel":"doctrine","datetime":"2024-08-18T12:11:49.548382+05:45","extra":{}}
{"message":"Rolling back transaction","context":{},"level":100,"level_name":"DEBUG","channel":"doctrine","datetime":"2024-08-18T12:11:49.548573+05:45","extra":{}}
{"message":"Uncaught PHP Exception Doctrine\DBAL\Driver\PDO\PDOException: "There is no active transaction" at /opt/www/vendor/doctrine/dbal/src/Driver/PDO/PDOException.php line 20","context":{"exception":{"class":"Doctrine\DBAL\Driver\PDO\PDOException","message":"There is no active transaction","code":0,"file":"/opt/www/vendor/doctrine/dbal/src/Driver/PDO/PDOException.php:20","previous":{"class":"PDOException","message":"There is no active transaction","code":0,"file":"/opt/www/vendor/doctrine/dbal/src/Driver/PDO/Connection.php:135"}},"errors":null},"level":500,"level_name":"CRITICAL","channel":"app","datetime":"2024-08-18T12:11:49.548958+05:45","extra":{}}

i already asked this 
https://github.com/doctrine/DoctrineBundle/issues/1815
https://github.com/dunglas/frankenphp/issues/978
they suggest me to your commit
still i am not able to solve my issue
